cip36/src/lib.rs

```diff
@@ -0,0 +1,602 @@
+    #[error("Invalid delegation weights")]
+    DelegationWeightsZero,
+}
+
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub enum DelegationDistribution {
+    Weighted {
+        delegations: Vec<Delegation>,
+        #[serde(skip)]
+        delegations_encoding: LenEncoding,
+    },
+    LegacyKeyRegistration(LegacyKeyRegistration),
+}
+
+impl DelegationDistribution {
+    /// Create a new weighted delegation. Weights are relative to all others and will be rounded down.
+    /// Leftover ADA will be delegated to the last item in the array.
+    pub fn new_weighted(delegations: Vec<Delegation>) -> Self {
+        Self::Weighted {
+            delegations,
+            delegations_encoding: LenEncoding::default(),
+        }
+    }
+
+    /// Delegate to a single key i.e. CIP-15.
+    pub fn new_legacy_key_registration(legacy_key_registration: LegacyKeyRegistration) -> Self {
+        Self::LegacyKeyRegistration(legacy_key_registration)
+    }
+}
+
+/// Weighted delegation input.
+/// This is the proportion of weight to assign to this public key relative to the weights
+/// of all other Delegations where this is used.
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub struct Delegation {
+    pub voting_pub_key: VotingPubKey,
+    pub weight: Weight,
+    #[serde(skip)]
+    pub encodings: Option<DelegationEncoding>,
+}
+
+impl Delegation {
+    pub fn new(voting_pub_key: VotingPubKey, weight: Weight) -> Self {
+        Self {
+            voting_pub_key,
+            weight,
+            encodings: None,
+        }
+    }
+}
+
+/// This is the entire metadata schema for CIP-36 deregistration.
+/// It can be parsed by passing in the CBOR bytes of the entire transaction metadatum
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub struct DeregistrationCbor {
+    pub key_deregistration: KeyDeregistration,
+    pub deregistration_witness: DeregistrationWitness,
+}
+
+impl DeregistrationCbor {
+    pub fn new(key_deregistration: KeyDeregistration, deregistration_witness: DeregistrationWitness) -> Self {
+        Self {
+            key_deregistration,
+            deregistration_witness,
+        }
+    }
+
+    /// Add to an existing metadata (could be empty) the full CIP36 deregistration metadata
+    pub fn add_to_metadata(&self, metadata: &mut Metadata) -> Result<(), DeserializeError> {
+        let dereg_metadatum = TransactionMetadatum::from_cbor_bytes(&self.key_deregistration.to_original_cbor_bytes())?;
+        metadata.insert(KEY_DEREGISTRATION_LABEL, dereg_metadatum);
+        let witness_metadatum = TransactionMetadatum::from_cbor_bytes(&self.deregistration_witness.to_original_cbor_bytes())?;
+        metadata.insert(DEREGISTRATION_WITNESS_LABEL, witness_metadatum);
+        Ok(())
+    }
+
+    // these are not implementing Serialize/Deserialize as we do not keep track of the rest of the encoding metadata
```

Comment:
I didn't like them implementing the same `Serialize` and having it provide `to_original_cbor_bytes()` when that isn't quite true as it might be a subset. I don't think this solution is particularly pretty but it should lead to less possible confusion. You really shouldn't be round-tripping this anyway, but instead using it for either:
1) creating the metadata in the first place
2) parsing on-chain stuff for read-only purposes that came from `Metadata` at some point (use the `Metadata` object for the round-tripping if you care about encodings)